File: benchmarks/python/harness.py
"""Benchmark harness for PyTorch — mirrors the Rust harness."""

import json
import resource
import statistics
import sys
import time

import torch


def get_device():
    if torch.cuda.is_available():
        torch.backends.cudnn.benchmark = True
        dev = torch.device("cuda")
        name = torch.cuda.get_device_name(0)
        print(f"device: CUDA ({name})", file=sys.stderr)
    else:
        dev = torch.device("cpu")
        print("device: CPU", file=sys.stderr)
    print(file=sys.stderr)
    return dev


def run_benchmark(name, model, batches, device, optimizer,
                  loss_fn=torch.nn.MSELoss(),
                  warmup_epochs=3, measured_epochs=20):
    """Run a benchmark and return a result dict."""
    param_count = sum(p.numel() for p in model.parameters())

    # Reset VRAM tracking before each benchmark
    if device.type == "cuda":
        torch.cuda.empty_cache()
        torch.cuda.reset_peak_memory_stats()

    # Warmup
    for _ in range(warmup_epochs):
        for x, y in batches:
            pred = model(x)
            loss = loss_fn(pred, y)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

    if device.type == "cuda":
        torch.cuda.synchronize()

    # Measured epochs
    epoch_times = []
    final_loss = 0.0

    for _ in range(measured_epochs):
        start = time.perf_counter()
        epoch_loss = 0.0

        for x, y in batches:
            pred = model(x)
            loss = loss_fn(pred, y)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            epoch_loss += loss.item()

        if device.type == "cuda":
            torch.cuda.synchronize()

        elapsed_ms = (time.perf_counter() - start) * 1000
        epoch_times.append(elapsed_ms)
        final_loss = epoch_loss / len(batches)

    # Stats
    sorted_times = sorted(epoch_times)
    median = statistics.median(epoch_times)
    mean = statistics.mean(epoch_times)

    # VRAM — report both allocated (active tensors) and reserved (allocator pool)
    vram_mb = None
    vram_reserved_mb = None
    if device.type == "cuda":
        vram_mb = torch.cuda.max_memory_allocated() / (1024 * 1024)
        vram_reserved_mb = torch.cuda.max_memory_reserved() / (1024 * 1024)

    # RSS
    rss_mb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024  # Linux: kB

    return {
        "name": name,
        "device": f"CUDA ({torch.cuda.get_device_name(0)})" if device.type == "cuda" else "CPU",
        "warmup_epochs": warmup_epochs,
        "measured_epochs": measured_epochs,
        "batches_per_epoch": len(batches),
        "batch_size": batches[0][0].shape[0],
        "param_count": param_count,
        "epoch_times_ms": epoch_times,
        "median_epoch_ms": median,
        "mean_epoch_ms": mean,
        "min_epoch_ms": sorted_times[0],
        "max_epoch_ms": sorted_times[-1],
        "final_loss": final_loss,
        "vram_mb": vram_mb,
        "vram_reserved_mb": vram_reserved_mb,
        "rss_mb": rss_mb,
    }


def print_result(r):
    print(f"  {r['name']}", file=sys.stderr)
    print(f"    device:     {r['device']}", file=sys.stderr)
    print(f"    params:     {format_count(r['param_count'])}", file=sys.stderr)
    print(f"    batches:    {r['batches_per_epoch']} x {r['batch_size']}", file=sys.stderr)
    print(f"    epochs:     {r['warmup_epochs']} warmup + {r['measured_epochs']} measured", file=sys.stderr)
    print(f"    median:     {r['median_epoch_ms']:.1f} ms/epoch", file=sys.stderr)
    print(f"    mean:       {r['mean_epoch_ms']:.1f} ms/epoch", file=sys.stderr)
    print(f"    range:      {r['min_epoch_ms']:.1f} - {r['max_epoch_ms']:.1f} ms", file=sys.stderr)
    print(f"    final loss: {r['final_loss']:.6f}", file=sys.stderr)
    if r["vram_mb"] is not None:
        print(f"    VRAM alloc: {r['vram_mb']:.0f} MB", file=sys.stderr)
        print(f"    VRAM rsrvd: {r['vram_reserved_mb']:.0f} MB", file=sys.stderr)
    print(f"    RSS:        {r['rss_mb']:.0f} MB", file=sys.stderr)
    print(file=sys.stderr)


def format_count(n):
    if n >= 1_000_000:
        return f"{n / 1_000_000:.1f}M"
    elif n >= 1_000:
        return f"{n / 1_000:.1f}K"
    return str(n)


def print_summary(results):
    print("=== Summary ===", file=sys.stderr)
    print(file=sys.stderr)
    print(f"  {'benchmark':<20} {'median':>10} {'mean':>10} {'params':>10} {'alloc':>10} {'reserved':>10}", file=sys.stderr)
    print(f"  {'-' * 74}", file=sys.stderr)
    for r in results:
        alloc = f"{r['vram_mb']:.0f} MB" if r["vram_mb"] is not None else "—"
        rsrvd = f"{r['vram_reserved_mb']:.0f} MB" if r.get("vram_reserved_mb") is not None else "—"
        print(
            f"  {r['name']:<20} {r['median_epoch_ms']:>8.1f}ms {r['mean_epoch_ms']:>8.1f}ms "
            f"{format_count(r['param_count']):>10} {alloc:>10} {rsrvd:>10}",
            file=sys.stderr,
        )
    print(file=sys.stderr)

File: benchmarks/src/harness.rs
//! Benchmark harness: timing, warmup, VRAM sampling, JSON output.

use std::time::Instant;

use serde::Serialize;

/// Result of a single benchmark run.
#[derive(Debug, Serialize)]
pub struct BenchResult {
    pub name: String,
    pub device: String,
    pub warmup_epochs: usize,
    pub measured_epochs: usize,
    pub batches_per_epoch: usize,
    pub batch_size: usize,
    pub param_count: usize,
    pub epoch_times_ms: Vec<f64>,
    pub median_epoch_ms: f64,
    pub mean_epoch_ms: f64,
    pub min_epoch_ms: f64,
    pub max_epoch_ms: f64,
    pub final_loss: f64,
    /// Active tensor bytes (matches torch.cuda.memory_allocated semantics).
    pub vram_mb: Option<f64>,
    /// Total allocator reservation (matches torch.cuda.memory_reserved semantics).
    pub vram_reserved_mb: Option<f64>,
    pub rss_mb: f64,
}

/// Configuration for running a benchmark.
pub struct BenchConfig {
    pub name: String,
    pub warmup_epochs: usize,
    pub measured_epochs: usize,
    pub batches_per_epoch: usize,
    pub batch_size: usize,
    /// Active VRAM (bytes) before benchmark setup.
    pub vram_baseline: u64,
    /// Reserved VRAM (bytes) before benchmark setup.
    pub vram_reserved_baseline: u64,
}

impl Default for BenchConfig {
    fn default() -> Self {
        Self {
            name: String::new(),
            warmup_epochs: 3,
            measured_epochs: 20,
            batches_per_epoch: 100,
            batch_size: 128,
            vram_baseline: 0,
            vram_reserved_baseline: 0,
        }
    }
}

/// Flush the caching allocator and snapshot VRAM baselines (active, reserved).
/// Returns `(0, 0)` on CPU.
pub fn vram_baseline() -> (u64, u64) {
    if flodl::cuda_available() {
        flodl::cuda_empty_cache();
        let active = flodl::cuda_active_bytes().unwrap_or(0);
        let reserved = flodl::cuda_allocated_bytes().unwrap_or(0);
        (active, reserved)
    } else {
        (0, 0)
    }
}

/// Run a benchmark with the given configuration.
///
/// `run_epoch` is called for each epoch and should return the average loss.
/// It receives `(epoch_index, is_warmup)`.
pub fn run_benchmark(
    config: &BenchConfig,
    param_count: usize,
    mut run_epoch: impl FnMut(usize, bool) -> flodl::Result<f64>,
) -> flodl::Result<BenchResult> {
    let device_name = if flodl::cuda_available() {
        format!("CUDA ({})", flodl::cuda_device_name().unwrap_or_default())
    } else {
        "CPU".to_string()
    };

    // Warmup
    for i in 0..config.warmup_epochs {
        run_epoch(i, true)?;
    }

    // Sync before measurement
    #[cfg(feature = "cuda")]
    flodl::cuda_synchronize(0);

    // Measured epochs
    let mut epoch_times = Vec::with_capacity(config.measured_epochs);
    let mut final_loss = 0.0;

    for i in 0..config.measured_epochs {
        let start = Instant::now();
        final_loss = run_epoch(i, false)?;

        #[cfg(feature = "cuda")]
        flodl::cuda_synchronize(0);

        epoch_times.push(start.elapsed().as_secs_f64() * 1000.0);
    }

    // VRAM — active tensors (matches torch.cuda.memory_allocated)
    let vram_mb = flodl::cuda_active_bytes()
        .ok()
        .map(|used| {
            let delta = used.saturating_sub(config.vram_baseline);
            delta as f64 / (1024.0 * 1024.0)
        });

    // VRAM — allocator reservation (matches torch.cuda.memory_reserved)
    let vram_reserved_mb = flodl::cuda_allocated_bytes()
        .ok()
        .map(|used| {
            let delta = used.saturating_sub(config.vram_reserved_baseline);
            delta as f64 / (1024.0 * 1024.0)
        });

    // RSS
    let rss_mb = flodl::rss_kb() as f64 / 1024.0;

    // Stats
    let mut sorted = epoch_times.clone();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());
    let median = sorted[sorted.len() / 2];
    let mean = epoch_times.iter().sum::<f64>() / epoch_times.len() as f64;
    let min = sorted[0];
    let max = sorted[sorted.len() - 1];

    Ok(BenchResult {
        name: config.name.clone(),
        device: device_name,
        warmup_epochs: config.warmup_epochs,
        measured_epochs: config.measured_epochs,
        batches_per_epoch: config.batches_per_epoch,
        batch_size: config.batch_size,
        param_count,
        epoch_times_ms: epoch_times,
        median_epoch_ms: median,
        mean_epoch_ms: mean,
        min_epoch_ms: min,
        max_epoch_ms: max,
        final_loss,
        vram_mb,
        vram_reserved_mb,
        rss_mb,
    })
}

/// Pretty-print a benchmark result to stderr.
pub fn print_result(r: &BenchResult) {
    eprintln!("  {}", r.name);
    eprintln!("    device:     {}", r.device);
    eprintln!("    params:     {}", format_count(r.param_count));
    eprintln!("    batches:    {} x {}", r.batches_per_epoch, r.batch_size);
    eprintln!("    epochs:     {} warmup + {} measured", r.warmup_epochs, r.measured_epochs);
    eprintln!("    median:     {:.1} ms/epoch", r.median_epoch_ms);
    eprintln!("    mean:       {:.1} ms/epoch", r.mean_epoch_ms);
    eprintln!("    range:      {:.1} - {:.1} ms", r.min_epoch_ms, r.max_epoch_ms);
    eprintln!("    final loss: {:.6}", r.final_loss);
    if let Some(vram) = r.vram_mb {
        eprintln!("    VRAM alloc: {:.0} MB", vram);
    }
    if let Some(vram) = r.vram_reserved_mb {
        eprintln!("    VRAM rsrvd: {:.0} MB", vram);
    }
    eprintln!("    RSS:        {:.0} MB", r.rss_mb);
    eprintln!();
}

fn format_count(n: usize) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.1}K", n as f64 / 1_000.0)
    } else {
        format!("{n}")
    }
}
